### PR TITLE
chore(build): add concurrency to the github workflow to conditionally queue builds.

### PR DESCRIPTION
**Pull request by vegano1**
# Overview

Closes: [PLAT-309](https://opentrons.atlassian.net/browse/PLAT-309)

We make a lot of builds, which is great for having a pulse on the system. However, this leads to excessive builds that are never used, costing us computation time and money. This pull request fixes this problem (or greatly mitigates it) by adding Github [concurrency](https://docs.github.com/en/actions/using-jobs/using-concurrency) to the workflow, which cancels queued builds with the same signature. The build signature is just a concatenated string of the monorepo, buildroot refs, plus build variant and build-type used when building the image. Here are some examples,
```
'refs/heads/edge refs/heads/opentrons-develop internal-release release'
'refs/tags/v7.3.0-alpha.2 refs/tags/v0.6.4 refs/tags/v50 release release'
```
This means that a build will start right away if there are no other builds in progress; when another build with the same signature is added, it will be queued as they do now. However, if we now have a build in-progress, a build queued, and a new build with the same signature is added to the queue, the currently queued build will be canceled, as seen in the image below. So at any given time, we can have one in-progress build and, at most, one queued build of the same signature.
Screen Shot 2024-05-15 at 10 29 43 AM

[PLAT-309]: https://opentrons.atlassian.net/browse/PLAT-309?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -56,6 +56,9 @@ jobs:
     name: 'Building ${{needs.decide-refs.outputs.variant}} images on ${{ matrix.build_env }}'
     timeout-minutes: 480
     runs-on: ['self-hosted', '${{matrix.build_env}}', '${{needs.decide-refs.outputs.variant}}']
+    concurrency:
+      group: ${{needs.decide-refs.outputs.monorepo}} ${{needs.decide-refs.outputs.buildroot}} ${{needs.decide-refs.outputs.variant}} ${{needs.decide-refs.outputs.build-type}}
```
Comment (SyntaxColoring) [inline .github/workflows/build.yml]: Do we need `inputs.infra-stage` in here?

Comment (vegano1) [reply .github/workflows/build.yml]: Don't see why not, good call!

Comment (vegano1) [reply .github/workflows/build.yml]: done

**Review by SyntaxColoring — APPROVED** [review]
Dope.

**Pull request merged**